flink-runtime/src/main/java/org/apache/flink/runtime/concurrent/FutureUtils.java

```diff
@@ -802,6 +804,89 @@ public void onComplete(Throwable failure, U success) {
 		return result;
 	}
 
+	/**
+	 * This function takes a {@link CompletableFuture} and a function to apply to this future. If the input future
+	 * is already done, this function returns {@link CompletableFuture#thenApply(Function)}. Otherwise, the return
+	 * value is {@link CompletableFuture#thenApplyAsync(Function, Executor)} with the given executor.
+	 *
+	 * @param completableFuture the completable future for which we want to apply.
+	 * @param executor the executor to run the apply function if the future is not yet done.
+	 * @param applyFun the function to apply.
+	 * @param <IN> type of the input future.
+	 * @param <OUT> type of the output future.
+	 * @return a completable future that is applying the given function to the input future.
+	 */
+	public static <IN, OUT> CompletableFuture<OUT> applyAsyncIfNotDone(
+		CompletableFuture<IN> completableFuture,
+		Executor executor,
+		Function<? super IN, ? extends OUT> applyFun) {
+		return completableFuture.isDone() ?
```

Comment:
I think this should be fine. There is indeed a race condition between completing the future and calling `thenApplyAsync` but in the worst case you execute the callback in the supplied executor instead of the calling thread. Both paths should work.

Reply:
Till is correct, this is only an optimization but it is ok to run a future that is completed after the `isDone` check as async.